### DISCUSSION TITLE
On-demand ISR Not Working on Vercel

### DISCUSSION THREAD

**Discussion by dubbajubba** [vercel/next.js#42971] [Help] [answered]
Hi,

I'm having an issue with getting on-demand ISR to work when deployed to Vercel and I do not understand why.

My goal is to revalidate my pages every time a user updates the page. When a user updates their page, the update request is made to an update API route in my NextJS application. This route makes a call to an external API to update the record, then makes a request to the revalidate API endpoint. Here is piece of code that handles the request,
```typescript
try {
  const getRes = await update(id.toString(), req.body);

  if (!getRes) throw Error('Could not update');

  if (req.body.slug) {
    fetch(
      `${process.env.NEXT_PUBLIC_URL}/api/revalidate?secret=${process.env.REVALIDATE_SECRET_TOKEN}&revalidateUrl=/q/${req.body.slug}`,
    );
  }

  return res.status(204).end('Updated');
} catch (err: any) {
  res
    .status(500)
    .json({ error: { statusCode: 500, message: err.message } });
}
```

Here is my revalidate.ts file,
```typescript
import { NextApiRequest, NextApiResponse } from 'next';

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse,
) {
  // Check for secret to confirm this is a valid request
  if (req.query.secret !== process.env.REVALIDATE_SECRET_TOKEN) {
    return res.status(401).json({ message: 'Invalid token' });
  }

  try {
    const { revalidateUrl } = req.query;
    // this should be the actual path not a rewritten path
    // e.g. for "/blog/[slug]" this should be "/blog/post-1"
    if (revalidateUrl && typeof revalidateUrl === 'string') {
      await res.revalidate(revalidateUrl);
      return res.json({ revalidated: true });
    }

    return res.status(500).send('Error revalidating');
  } catch (err) {
    // If there was an error, Next.js will continue
    // to show the last successfully generated page
    return res.status(500).send('Error revalidating');
  }
}
```

A couple things to note:
- When I run `yarn build` and `yarn start` locally it works fine.
- In the deployed environment, if I manually type the revalidate URL into my browser and visit it, it revalidates the page. I can visit the page and confirm that it has been revalidated.
- I'm currently on NextJS version `12.3.3`

I have read through all the documentation I can find and can't see anything that I am doing wrong. Would love to find the solution to this, it's driving me crazy!

Thanks!

**Comment by ijjk** [comment]
Hi, why is `NEXT_PUBLIC_URL` being used here? The request should be made to the same hostname as the user is currently visiting the page as otherwise the cache will only be updated for the hostname specific to the `NEXT_PUBLIC_URL` value since the deployment cache is currently unique per hostname.

**Reply by dubbajubba** [reply]
Hey @ijjk, thanks for your response! I use that as the base URL throughout the site. The `NEXT_PUBLIC_URL` will always be the hostname that the user is visiting since I am using that variable to provide the correct hostname per environment.

Please let me know if I'm misunderstanding something here.

**Reply by ijjk** [reply]
Can you share a repo/deployment URL where you are seeing the revalidate not update correctly?

**Comment by icyJoseph** [comment]
Are you seeing any logs on the Vercel dashboard? 

I hope the error is not the missing `'` on your snippet.

I'd try to also differentiate the try/catch branches with different error messages, and possibly include some minimal debugging information. 

If it doesn't work when deploying then it is probably an expectation we have that's not really being fulfilled. 

Perhaps the API is never even reached to begin with? Or the token secret is being encoded, and thus the strings are not equal anymore. Like if the secret contains a `$` sign, or other reserved characters, they might be getting encoded to `%24` form, and then the equality check fails.

**Reply by dubbajubba** [reply]
Hi @icyJoseph, thanks for your response! Sorry about the missing `'` in my example. I had actually reduced the sentence in that response for this GitHub discussion, but the source code does have the character. I have updated my post and fixed it.

I do not get any errors in the Vercel dashboard. I will add some more granular debugging.

Based on the code I'm sharing, can you think of any expectation that Vercel may have that I am not fulfilling? I haven't found a ton of documentation on On-Demand ISR, however, from what I've found, it seems to me that I'm meeting the requirements.

My secret didn't contain any reserved characters, but I encoded and decoded it to be safe, but it still is not revalidating. I wouldn't see why the API wouldn't be reached in deployment, but would be hit when I do a production build locally.

**Reply by icyJoseph** [reply]
I am really not sure, this kind of situation where it works locally but not on deploy, and that show no obvious error, are difficult to trace, here's another person having a similar situation, https://github.com/vercel/next.js/discussions/42685.

Hopefully adding logs sheds more light into this.

**Reply by dubbajubba** [reply]
I just read through the discussion you linked and it actually seems like the same issue! I guess I didn't see that discussion before because it was specific about navigating with next/link, however, I can make the update on one browser/computer, and when I visit it in a different browser/computer, it is still showing the stale page.

**Comment by dubbajubba** [comment]
So after some more testing, I have found some weird behavior. 

First off, it seems the revalidation is actually working, the server just seems to be serving a cached version of the page. The interesting thing is, if I make a change to the page, and visit it on another computer or browser, it will load the previous (stale) version of the page. However, if I simply refresh the page, the latest version of the page with the update is now showing. In my vercel.json file, I have the route path for these pages set to `max-age=0` and `s-maxage=86400`. Here is what it looks like,
```
    {
      "source": "/q/[slug].ts",
      "headers": [
        {
          "key": "Cache-Control",
          "value": "max-age=0"
        },
        {
          "key": "Cache-Control",
          "value": "s-maxage=86400"
        }
      ]
    }
```

And here is a screenshot of the request's response headers, showing that the max-age=0, so it shouldn't be using browser cache.
![image](https://user-images.githubusercontent.com/33437206/202287178-d204a6c6-55ab-47dc-8be8-9fe1b1cf8586.png)

So for some reason Vercel is serving up a cached version of the page, even though the page has been revalidated and changed.

**Comment by dubbajubba** [comment]
I'm looking at the example on-demand ISR repo https://github.com/leerob/on-demand-isr and notice it is using React v18. I am still using React v17, I would think React v17 would still work, but wanted to check, React v17 should work with on-demand ISR, right?

**Comment by dubbajubba** [comment, marked as answer]
I solved my issue! I didn't think that I needed to await the GET request to the validate API endpoint. After adding await to that call, I'm consistently getting the latest version of the cached page.

Below is the updated block of code from my update.ts file.

```typescript
try {
  const getRes = await update(id.toString(), req.body);

  if (!getRes) throw Error('Could not update');

  if (req.body.slug) {
    await fetch(
      `${process.env.NEXT_PUBLIC_URL}/api/revalidate?secret=${process.env.REVALIDATE_SECRET_TOKEN}&revalidateUrl=/q/${req.body.slug}`,
    );
  }

  return res.status(204).end('Updated');
} catch (err: any) {
  res
    .status(500)
    .json({ error: { statusCode: 500, message: err.message } });
}
```

Only change is adding the `await` in front of `fetch()`

**Reply by icyJoseph** [reply]
Oh wow yeah, easy to miss! Glad you fixed it!

**Reply by dubbajubba** [reply]
@icyJoseph, I actually have a question, not sure if you know the answer. In some cases I want to make a call to an external API from my NextJS server. However, I don't want to await the response from the external API before returning a response to the user of my app. I assumed that the function would continue to complete on the server even if the user is not awaiting it. Is this not the case?

For example, I have a middleware where I want to record the User Agent data and POST it to my external API. The recording of this information is something I want to run in the background and don't want to slow my response to the user. If I don't await the POST, will it fail to POST if I return a response to the user before it completes?

**Reply by icyJoseph** [reply]
Once you create the fetch request, even if you ignore the returned promise, the request continues its lifecycle. Unless the process is being killed, then there might be race conditions between the time the sync code stops and whether or not the request has been started.

I had assumed your revalidate function was running in a whole different project, and that it called the Next.js API route to revalidate a page, and then that project went on to do other stuff. If you want to be sure revalidation is done, and then do something you gotta await the fetch request.

Regarding the last case,  once you return or resolve the response, the runtime might kill the serverless lambda or middleware. For example Vercel API routes are destroyed after 10 seconds IIRC. Some runtimes have mechanisms to queue work and tell to wait for its completion. Do remember that a service like Vercel for example, has maximum interest in releasing resources as soon as possible, to use them to serve another request.

In a sense what you want is a controlled memory leak, the serverless function/handler returns but there's a promise hanging. Extreme simplification follows, but normally if there's anything in the event loop, upon completing all of the sync code, the node engine process the event loop, unless something, either a call to process.exit, an unhandled rejection of the user sending a signal to kill the process, stops it.

**Reply by dubbajubba** [reply]
Thanks for the detailed information @icyJoseph, very helpful!
